{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Motif-Search\" data-toc-modified-id=\"Motif-Search-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Motif Search</a></span><ul class=\"toc-item\"><li><span><a href=\"#Pre-run-MEME\" data-toc-modified-id=\"Pre-run-MEME-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Pre-run MEME</a></span></li><li><span><a href=\"#Compare-motifs-to-RegulonDB-motifs\" data-toc-modified-id=\"Compare-motifs-to-RegulonDB-motifs-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Compare motifs to RegulonDB motifs</a></span></li></ul></li><li><span><a href=\"#Helper-functions\" data-toc-modified-id=\"Helper-functions-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Helper functions</a></span><ul class=\"toc-item\"><li><span><a href=\"#Fonts\" data-toc-modified-id=\"Fonts-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Fonts</a></span></li><li><span><a href=\"#Colors\" data-toc-modified-id=\"Colors-2.2\"><span class=\"toc-item-num\">2.2&nbsp;&nbsp;</span>Colors</a></span></li><li><span><a href=\"#Fitting\" data-toc-modified-id=\"Fitting-2.3\"><span class=\"toc-item-num\">2.3&nbsp;&nbsp;</span>Fitting</a></span></li><li><span><a href=\"#Misc\" data-toc-modified-id=\"Misc-2.4\"><span class=\"toc-item-num\">2.4&nbsp;&nbsp;</span>Misc</a></span></li></ul></li><li><span><a href=\"#Plot-functions\" data-toc-modified-id=\"Plot-functions-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Plot functions</a></span></li><li><span><a href=\"#Final-PDF-Maker\" data-toc-modified-id=\"Final-PDF-Maker-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Final PDF Maker</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from tqdm import tqdm_notebook as tqdm\n",
    "import matplotlib.gridspec as gridspec\n",
    "from PIL import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('../')\n",
    "from icaviz.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = '../data/precise_data/'\n",
    "GENE_DIR = '../data/annotation/'\n",
    "enrich = pd.read_csv(DATA_DIR+'curated_enrichments.csv')\n",
    "names = enrich['name'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data = load_data(X=DATA_DIR+'log_tpm.csv',\n",
    "                     S=DATA_DIR+'S.csv',\n",
    "                     A=DATA_DIR+'A.csv',\n",
    "                     metadata=DATA_DIR+'metadata.csv',\n",
    "                     annotation=GENE_DIR+'gene_info.csv',\n",
    "                     trn=GENE_DIR+'TRN.csv',\n",
    "                     fasta=GENE_DIR+'NC_000913.3.fasta',\n",
    "                     cutoff = 550,\n",
    "                     names=names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning)\n",
    "\n",
    "# Set various parameters\n",
    "plt.ioff()\n",
    "\n",
    "sns.set_style('ticks')\n",
    "\n",
    "params = {'mathtext.default': 'regular','font.family':'sans-serif',\n",
    "          'font.sans-serif':['Arial'],\n",
    "          'pdf.fonttype':42,'ps.fonttype':42,\n",
    "          'xtick.labelsize':7,'xtick.major.size':3,'xtick.direction':'in',\n",
    "          'ytick.labelsize':7,'ytick.major.size':3,'ytick.direction':'in'\n",
    "          }\n",
    "plt.rcParams.update(params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Motif Search"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pre-run MEME"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In case you need to re-run the directory, this pre-calculates all the information for MEME and TOMTOM so you only need to run this block once"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Force -> True if you are running this for the first time\n",
    "force = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sites = []\n",
    "motifs = []\n",
    "\n",
    "comp_names = enrich[enrich.Category!='Genomic Alterations'].name\n",
    "# Run once using 300 nts upstream\n",
    "for k in tqdm(comp_names):\n",
    "    DF_motif,DF_sites = find_motifs(ica_data,k,nmotifs=1,verbose=False,evt=1e-3,\n",
    "                                    force=force,upstream=300,downstream=0,maxw=30)\n",
    "    \n",
    "    # Save all found binding sites\n",
    "    if len(DF_sites) > 0:\n",
    "        tmp_sites = DF_sites[DF_sites.site_seq.notnull()]\n",
    "        tmp_sites.index = pd.MultiIndex.from_tuples([(k, x[1]) for x in tmp_sites.index])\n",
    "        sites.append(tmp_sites)\n",
    "\n",
    "found_motifs1 = [k for k,v in ica_data.motif_info.items() if len(v[0])>0]\n",
    "\n",
    "# Run again for remaining sequences with 600 nt upstream and 100 downstream\n",
    "for k in tqdm(set(comp_names) - set(found_motifs1)):\n",
    "    DF_motif,DF_sites = find_motifs(ica_data,k,nmotifs=1,verbose=False,evt=1e-3,\n",
    "                                    force=force,upstream=600,downstream=100,maxw=30)\n",
    "    \n",
    "    # Save all found binding sites\n",
    "    if len(DF_sites) > 0:\n",
    "        tmp_sites = DF_sites[DF_sites.site_seq.notnull()]\n",
    "        tmp_sites.index = pd.MultiIndex.from_tuples([(k, x[1]) for x in tmp_sites.index])\n",
    "        sites.append(tmp_sites)\n",
    "    \n",
    "found_motifs2 = [k for k,v in ica_data.motif_info.items() if len(v[0])>0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare motifs to RegulonDB motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "force = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = []\n",
    "motif_db = '../data/annotation/regulonDB_MEME.txt'\n",
    "\n",
    "for k in tqdm(found_motifs2):\n",
    "    DF_motif,_ = find_motifs(ica_data,k,verbose=False,force=False)\n",
    "    DF_motif.index = [k]\n",
    "    DF_motif['motif_img'] = ['motifs/'+re.sub('/','_',k)+'/logo1.eps']\n",
    "    _,compare_str,tomtom_img = compare_motifs(k,motif_db,evt=0.01,force=force)\n",
    "    DF_motif['tomtom_str'] = [compare_str]\n",
    "    DF_motif['tomtom_img'] = [tomtom_img]\n",
    "    dfs.append(DF_motif)\n",
    "        \n",
    "DF_motif_final = pd.concat(dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_motif_final.sort_values('e_value')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat(sites).to_csv('raw_motif_search.csv')\n",
    "DF_motif_final.to_csv('raw_motif_comparison.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Helper functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fonts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from fontTools.ttLib import TTFont\n",
    "font = TTFont('/usr/share/fonts/truetype/msttcorefonts/arial.ttf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "font_cmap = font['cmap']\n",
    "font_t = font_cmap.getcmap(3,1).cmap\n",
    "font_s = font.getGlyphSet()\n",
    "font_units_per_em = font['head'].unitsPerEm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from fontTools.ttLib import TTFont\n",
    "from fontTools.ttLib.tables._c_m_a_p import CmapSubtable\n",
    "\n",
    "def getTextWidth(text,pointSize):\n",
    "    total = 0\n",
    "    for c in text:\n",
    "        if ord(c) in font_t and font_t[ord(c)] in font_s:\n",
    "            total += font_s[font_t[ord(c)]].width\n",
    "        else:\n",
    "            total += font_s['.notdef'].width\n",
    "    total = total*float(pointSize)/font_units_per_em;\n",
    "    return (total/72,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def word_wrap(text, width, fontsize):\n",
    "    '''\n",
    "    Word wrap function / algorithm for wrapping text using proportional (versus \n",
    "    fixed-width) fonts.\n",
    "    \n",
    "    `text`: a string of text to wrap\n",
    "    `width`: the width in pixels to wrap to\n",
    "    `extent_func`: a function that returns a (w, h) tuple given any string, to\n",
    "                   specify the size (text extent) of the string when rendered. \n",
    "                   the algorithm only uses the width.\n",
    "    \n",
    "    Returns a list of strings, one for each line after wrapping.\n",
    "    '''\n",
    "    extent_func = lambda x: getTextWidth(x,fontsize)\n",
    "    lines = []\n",
    "    pattern = re.compile(r'(\\s+)')\n",
    "    lookup = dict((c, extent_func(c)[0]) for c in set(text))\n",
    "    for line in text.splitlines():\n",
    "        tokens = pattern.split(line)\n",
    "        tokens.append('')\n",
    "        widths = [sum(lookup[c] for c in token) for token in tokens]\n",
    "        start, total = 0, 0\n",
    "        for index in range(0, len(tokens), 2):\n",
    "            if total + widths[index] > width:\n",
    "                end = index + 2 if index == start else index\n",
    "                lines.append(''.join(tokens[start:end]))\n",
    "                start, total = end, 0\n",
    "                if end == index + 2:\n",
    "                    continue\n",
    "            total += widths[index] + widths[index + 1]\n",
    "        if start < len(tokens):\n",
    "            lines.append(''.join(tokens[start:]))\n",
    "    lines = [line.strip() for line in lines]\n",
    "    return lines or ['']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clamp(val, minimum=0, maximum=255):\n",
    "    if val < minimum:\n",
    "        return minimum\n",
    "    if val > maximum:\n",
    "        return maximum\n",
    "    return val\n",
    "\n",
    "def colorscale(hexstr, scalefactor):\n",
    "    \"\"\"\n",
    "    Scales a hex string by ``scalefactor``. Returns scaled hex string.\n",
    "\n",
    "    To darken the color, use a float value between 0 and 1.\n",
    "    To brighten the color, use a float value greater than 1.\n",
    "\n",
    "    >>> colorscale(\"#DF3C3C\", .5)\n",
    "    #6F1E1E\n",
    "    >>> colorscale(\"#52D24F\", 1.6)\n",
    "    #83FF7E\n",
    "    >>> colorscale(\"#4F75D2\", 1)\n",
    "    #4F75D2\n",
    "    \"\"\"\n",
    "\n",
    "    hexstr = hexstr.strip('#')\n",
    "\n",
    "    if scalefactor < 0 or len(hexstr) != 6:\n",
    "        return hexstr\n",
    "\n",
    "    r, g, b = int(hexstr[:2], 16), int(hexstr[2:4], 16), int(hexstr[4:], 16)\n",
    "\n",
    "    r = clamp(r * scalefactor)\n",
    "    g = clamp(g * scalefactor)\n",
    "    b = clamp(b * scalefactor)\n",
    "\n",
    "    return \"#{:02x}{:02x}{:02x}\".format(int(r), int(g), int(b))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rgb2hex = lambda x:'#{:02x}{:02x}{:02x}'.format(int(x[0]*255),int(x[1]*255),int(x[2]*255))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fitting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.optimize import curve_fit\n",
    "from sklearn.metrics import r2_score\n",
    "\n",
    "def broken_line(x, A, B, C): # this is your 'straight line' y=f(x)\n",
    "    y = np.zeros(len(x),dtype=np.float)\n",
    "    y += (A*x+B) * (x >= C)\n",
    "    y += (A*C + B) * (x < C)\n",
    "    return y\n",
    "\n",
    "def solid_line(x,A,B):\n",
    "    y = (A*x+B)\n",
    "    return y\n",
    "\n",
    "def get_fit(x,y):\n",
    "    \n",
    "    def adj_r2(f,x,y,params):\n",
    "        n = len(x)\n",
    "        k = len(params)-1\n",
    "        r2 = r2_score(y,f(x,*params))\n",
    "        return 1 - np.true_divide((1-r2)*(n-1),(n-k-1))\n",
    "\n",
    "    all_params = []\n",
    "    for c in [min(x),np.mean(x),max(x)]:\n",
    "        try:\n",
    "            all_params.append(curve_fit(broken_line, x, y,p0=[1,1,c])[0])\n",
    "        except:\n",
    "            pass\n",
    "        \n",
    "    all_params.append(curve_fit(solid_line,x,y)[0])\n",
    "\n",
    "    best_r2 = -np.inf\n",
    "    for params in all_params:\n",
    "        if len(params) == 2:\n",
    "            r2 = adj_r2(solid_line,x,y,params)\n",
    "        else:\n",
    "            r2 = adj_r2(broken_line,x,y,params)\n",
    "            \n",
    "        if r2 > best_r2:\n",
    "            best_r2 = r2\n",
    "            best_params = params\n",
    "            \n",
    "    if best_r2 < 0:\n",
    "        return [0,np.mean(y)],0\n",
    "\n",
    "    return best_params,best_r2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Misc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "flatten = lambda l: [item for sublist in l for item in sublist]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reformat_label(text):\n",
    "    name,genes = text.split(':')\n",
    "    new_genes = re.sub(r'( |\\n)(.*?)(?:,|$)',r',\\g<1>$\\mathit{{\\g<2>}}$',genes,flags=re.M)[2:]\n",
    "    return r'{}: {}'.format(name,new_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_genes(ica_data,k,ax):\n",
    "    cutoff = ica_data.thresholds[k]\n",
    "    colors = [ica_data.gene_colors[gene] for gene in ica_data.S.index]\n",
    "    \n",
    "    # Draw scatterplot\n",
    "    baseline = ica_data.X[['control__wt_glc__1','control__wt_glc__2']].mean(axis=1)\n",
    "    scatter = ax.scatter(baseline.values,\n",
    "                         ica_data.S[k].values,\n",
    "                         c=colors,s=10,\n",
    "                         alpha=0.7,linewidth=0.0)\n",
    "    \n",
    "    ax.set_xlabel('Baseline Expression (log-TPM)',fontsize=10)\n",
    "    ax.set_ylabel('I-Modulon Gene Coefficients',fontsize=10)\n",
    " \n",
    "    # Get axes bounds\n",
    "    xmin,xmax = ax.get_xlim()\n",
    "    ymin,ymax = ax.get_ylim()\n",
    "            \n",
    "    # Add labels on datapoints        \n",
    "    component_genes = ica_data.show_enriched(k).index\n",
    "    texts = []\n",
    "    expand_args = {'expand_objects':(1.2,1.4),\n",
    "                   'expand_points':(1.3,1.3)}\n",
    "    \n",
    "    ## Put gene name if component contains under 20 genes\n",
    "    if len(component_genes) <= 25:\n",
    "        for gene in component_genes:\n",
    "            texts.append(ax.text(baseline[gene],\n",
    "                                 ica_data.S.loc[gene,k],\n",
    "                                 r'${}$'.format(ica_data.num2name[gene]),\n",
    "                                 fontsize=7,fontstyle='italic'))\n",
    "        expand_args['expand_text'] =(1.4,1.4)\n",
    "                                 \n",
    "    ## Repel texts from other text and points\n",
    "    rect = ax.add_patch(Rectangle((xmin,-cutoff),xmax-xmin,2*cutoff,fill=False,\n",
    "                     linewidth=0))\n",
    "        \n",
    "    adjust_text(texts,ax=ax,add_objects = [rect],\n",
    "                arrowprops=dict(arrowstyle=\"-\",color='k',lw=0.5),\n",
    "                only_move={'objects':'y'},**expand_args)\n",
    "    \n",
    "    # Draw horizontal dashed lines\n",
    "    ax.hlines([cutoff,-cutoff],xmin,xmax,colors='gray',linestyles='dashed',\n",
    "              linewidth=0.5)\n",
    "                                 \n",
    "\n",
    "    # Add legend\n",
    "    legend_info = []\n",
    "    num_lines = []\n",
    "    leg_fontsize = 8\n",
    "    for name,group in ica_data.gene_info.groupby('cog'):\n",
    "        # Get number of genes in COG\n",
    "        cog_genes = group.index\n",
    "        comp_cog_genes = set(ica_data.show_enriched(k).index) & set(cog_genes)\n",
    "\n",
    "        if len(comp_cog_genes) > 0:\n",
    "            gene_names = sorted([ica_data.num2name[gene] for gene in comp_cog_genes])\n",
    "            \n",
    "            # Create legend entry and wrap text\n",
    "            entry = '{} ({:d}): {}'.format(name,len(gene_names),', '.join(gene_names))\n",
    "            wrapped = word_wrap(entry,3,leg_fontsize)\n",
    "            text = '\\n'.join(wrapped)\n",
    "            num_lines.append(len(wrapped))\n",
    "            \n",
    "            # Create legend patch and add to list\n",
    "            legend_info.append((ica_data.cog_colors[name],text,len(gene_names),gene_names))\n",
    "\n",
    "    ## Sort legend by number of genes, keeping unknown function and no COG annotation at bottom\n",
    "    legend_info = sorted(legend_info,key=lambda x: (x[1][:10] not in ['Function u','No COG Ann'],x[2]),reverse=True)\n",
    "\n",
    "    max_lines = 10\n",
    "    ## If no legend entries, return\n",
    "    if len(legend_info) == 0:\n",
    "        return ax\n",
    "    ## If over X legend lines, only print first X-1 plus \"Other\"\n",
    "    elif sum(num_lines) > max_lines:\n",
    "        max_entries = max(np.where(np.cumsum(num_lines) < max_lines)[0])\n",
    "        other_genes = flatten([row[3] for row in legend_info[max_entries:]])\n",
    "        num_genes = len(other_genes)\n",
    "        if num_genes > 20:\n",
    "            other_entry = word_wrap('Other ({:d}): {} +{:d}'.format(num_genes,', '.join(other_genes[:20]),num_genes-20),3.5,leg_fontsize)\n",
    "        else:\n",
    "            other_entry = word_wrap('Other ({:d}): {}'.format(num_genes,', '.join(other_genes)),3.5,leg_fontsize)\n",
    "        other_text = '\\n'.join(other_entry)\n",
    "        legend_info = legend_info[:max_entries] + [('white',other_text,num_genes,other_genes)]\n",
    "\n",
    "    ## Add legend entries to plot as empty    \n",
    "    for info in legend_info:   \n",
    "        ax.plot([],[],\"o\",color=info[0],markersize=5,label=reformat_label(info[1]))\n",
    "\n",
    "    leg = ax.legend(loc='upper left', bbox_to_anchor=(1, 1.01),\n",
    "                    fontsize=leg_fontsize,title='COG Categories')\n",
    "\n",
    "    leg._legend_box.align = \"left\"\n",
    "    leg.get_title().set_fontweight('bold')\n",
    "    leg.get_title().set_fontsize(10)\n",
    "    \n",
    "    return ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_samples_bar(ica_data,k,ax):\n",
    "    \n",
    "    # Define sort function for conditions\n",
    "    def keyfxn(x):\n",
    "        name = x[0]\n",
    "        match = re.match('^.+?ale(\\d+)$',name)\n",
    "        if match:\n",
    "            return '{:02d}'.format(int(match.group(1)))\n",
    "        elif name.startswith('wt') or name.startswith('glu4'):\n",
    "            return '00'+name[3:]\n",
    "        else:\n",
    "            return name\n",
    "\n",
    "    # Set order for display\n",
    "    proj_order = ica_data.metadata.project_id.drop_duplicates()\n",
    "    rename = {'control':'','fur':'Iron','acid':'Acid\\n','oxidative':'Oxid.',\n",
    "              'nac_ntrc':'Nac/\\nNtrC','ompr':'OmpR\\n','misc':'Misc','omics':'Omics',\n",
    "              'minspan':'MinSpan','cra_crp':'Cra/\\nCrp','rpoB':'RpoB\\nKnock-in',\n",
    "              'crp':'Crp\\nARs','glu':'Glucose\\nEvol.',\n",
    "              '42c':'42C\\nEvol.','ssw':'Substrate-\\nswitching Evol.',\n",
    "              'pgi':'PGI KO\\nEvol.','ica':'Current\\nStudy','ytf':'Unchar.\\nTFs',\n",
    "              'efeU':'Pseudogene\\nRepair','pal':'Enz.\\nPromisc.','fps':'False\\nPos.'}\n",
    "    \n",
    "    \n",
    "    list2struct = []\n",
    "    for proj in proj_order:\n",
    "        group1 = ica_data.metadata[ica_data.metadata.project_id == proj]\n",
    "        for cond,group2 in sorted(group1.groupby('condition_id'),key=keyfxn):\n",
    "            for name in group2.index:\n",
    "                list2struct.append([proj,cond,ica_data.A.loc[k,name],len(group2)])\n",
    "    DF_comp = pd.DataFrame(list2struct,columns = ['project_id','condition_id','value','length'])\n",
    "    DF_comp['width'] = 1/DF_comp.length\n",
    "    DF_comp.index = [0]+np.cumsum(DF_comp.width).tolist()[:-1]\n",
    "\n",
    "    # Get xlabels and tick marks\n",
    "    xticks = []\n",
    "    xticklabels = []\n",
    "    vlines = []\n",
    "    proj_labels = []\n",
    "    proj_locs = []\n",
    "    for proj,group1 in DF_comp.groupby('project_id'):\n",
    "        vlines.append(min(group1.index))\n",
    "        proj_labels.append(rename[proj])\n",
    "        proj_locs.append(min(group1.index)+group1.width.sum()/2)\n",
    "        for cond,group2 in group1.groupby('condition_id'):\n",
    "            xticks.append(np.mean(group2.index)+np.mean(group2.width)/2)\n",
    "            xticklabels.append(cond)\n",
    "    \n",
    "    # Add alternating gray rectangles in back\n",
    "    for i in np.arange(0,len(DF_comp),2):\n",
    "        ax.add_patch(Rectangle((i,-100),1,200,color='#888888',alpha=0.1))\n",
    "    \n",
    "    \n",
    "    # Draw bar chart\n",
    "    ax.bar(DF_comp.index,DF_comp.value,width = DF_comp.width,align='edge',linewidth=0)\n",
    "    \n",
    "    # Set x axis parameters\n",
    "    ax.set_xlim([0,len(DF_comp)])\n",
    "    ax.set_xticks(xticks)\n",
    "    ax.set_xticklabels(xticklabels,fontsize=4)\n",
    "    ax.tick_params(axis='x',which='major',direction='in',length=4,labelrotation=90)\n",
    "    ax.tick_params(axis='y',labelsize=7)\n",
    "    ax.grid(False,which='major')\n",
    "    \n",
    "    # Add vertical bars to designate project\n",
    "    ax.vlines(vlines,-100,100,colors='#888888',linewidth=1)\n",
    "    ax.set_ylim(min(DF_comp.value)*1.2,max(DF_comp.value)*1.2)\n",
    "    ax.set_xlim(DF_comp.index.min(),DF_comp.index.max()+DF_comp.width.iloc[-1])\n",
    "    ax.set_ylabel('I-Modulon Activity',fontsize=6)\n",
    "\n",
    "\n",
    "    # Add project labels\n",
    "    ax2 = ax.twiny()\n",
    "    ax2.grid(False)\n",
    "    ax2.set_xlim(ax.get_xlim())\n",
    "    ax2.set_xticks(proj_locs)\n",
    "    ax2.set_xticklabels(proj_labels,fontsize=7)\n",
    "    ax2.tick_params(axis='x',length=0)\n",
    "    ax2.xaxis.tick_top()\n",
    "    \n",
    "    # Add top and bottom labels\n",
    "    ax.set_xlabel('Condition\\nID',x=-0.05,va='bottom',fontsize=9,labelpad=-5)\n",
    "    ax2.set_xlabel('Study',x=-0.05,va='top',fontsize=9,labelpad=3)\n",
    "    \n",
    "    return ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_histogram(ica_data,k,tfs,tf_info,ax):\n",
    "    pal = [rgb2hex(x) for x in sns.color_palette('Set1')]\n",
    "    pal = [pal[1],pal[2],pal[0]]+pal[3:]\n",
    "        \n",
    "    DF_gene = ica_data.component_DF(k,tfs=tfs)\n",
    "    bins = 20\n",
    "    \n",
    "    # Compute optimal range for histogram\n",
    "    xmin = min(min(DF_gene.comp),-ica_data.thresholds[k])\n",
    "    xmax = max(max(DF_gene.comp),ica_data.thresholds[k])\n",
    "    width = 2*ica_data.thresholds[k]/(np.floor(2*ica_data.thresholds[k]*bins/(xmax-xmin)-1))\n",
    "    xmin = -ica_data.thresholds[k]-width*np.ceil((-ica_data.thresholds[k] - xmin)/width)\n",
    "    xmax = xmin + width*bins\n",
    "        \n",
    "    ax.set_yscale('log', nonposy='clip')\n",
    "    ax.xaxis.grid(False)\n",
    "    ax.set_xlabel('I-Modulon Gene Coefficients',fontsize=10)\n",
    "    ax.set_ylabel('Count (log scale)',fontsize=10)\n",
    "    \n",
    "    # Plot histograms\n",
    "    if len(tfs) == 0:\n",
    "        label = 'No enriched\\nregulator'\n",
    "        color = '#888888'\n",
    "        \n",
    "        DF_gene['color'] = color\n",
    "        \n",
    "        ax.hist(DF_gene.comp,bins=bins,alpha=0.5,\n",
    "                range=(xmin,xmax),label=label,color=color,linewidth=0)\n",
    "        \n",
    "    elif len(tfs) == 1:\n",
    "        tf = tfs[0]\n",
    "        \n",
    "        # Plot unregulated genes\n",
    "        unreg_genes = DF_gene[~DF_gene[tf]].index\n",
    "        unreg_label = 'Unregulated\\nby {}'.format(tf[0].upper()+tf[1:])\n",
    "        unreg_color = '#aaaaaa'\n",
    "        DF_gene.loc[unreg_genes,'color'] = unreg_color\n",
    "        \n",
    "        ax.hist(DF_gene.loc[unreg_genes].comp,alpha=0.5,bins=bins,\n",
    "            range=(xmin,xmax),label=unreg_label,color=unreg_color,linewidth=0)\n",
    "        \n",
    "        # Plot regulated genes\n",
    "        reg_genes = DF_gene[DF_gene[tf]].index\n",
    "        reg_label = 'Regulated\\nby {}'.format(tf[0].upper()+tf[1:])\n",
    "        reg_color = pal[0]\n",
    "        DF_gene.loc[reg_genes,'color'] = reg_color\n",
    "        \n",
    "        ax.hist(DF_gene.loc[reg_genes].comp,alpha=0.7,bins=bins,\n",
    "            range=(xmin,xmax), label=reg_label,color=reg_color,linewidth=0)\n",
    "        \n",
    "    else:\n",
    "        # Plot unregulated genes\n",
    "        unreg_genes = DF_gene[~reduce(lambda x,y: (x | y),[DF_gene[tf] for tf in tfs])].index\n",
    "        unreg_label = 'Unregulated\\nby any'\n",
    "        unreg_color = '#aaaaaa'\n",
    "        DF_gene.loc[unreg_genes,'color'] = unreg_color\n",
    "        \n",
    "        ax.hist(DF_gene.loc[unreg_genes].comp,alpha=0.5,bins=bins,\n",
    "            range=(xmin,xmax),label=unreg_label,color=unreg_color,linewidth=0)        \n",
    "        \n",
    "        # Plot multiple regulated genes\n",
    "        multireg_genes = set()\n",
    "        for tf1,tf2 in combinations(tfs,2):\n",
    "            multireg_genes = multireg_genes.union(set(DF_gene[DF_gene[tf1] & DF_gene[tf2]].index))\n",
    "        multireg_label = 'Regulated by both' if len(tfs) ==2 else 'Regulated\\nby multiple'\n",
    "        multireg_color = '#8B4513'\n",
    "        DF_gene.loc[multireg_genes,'color'] = multireg_color\n",
    "        \n",
    "        # Get individually regulated genes\n",
    "        reg_genes = []\n",
    "        for i,tf in enumerate(tfs):\n",
    "            reg_genes = set(DF_gene[DF_gene[tf]].index) - set(multireg_genes)\n",
    "            reg_label = 'Regulated by\\nonly {}'.format(tf[0].upper()+tf[1:])\n",
    "            reg_color = pal[i]\n",
    "            DF_gene.loc[reg_genes,'color'] = reg_color\n",
    "            \n",
    "            if len(reg_genes) > 0:\n",
    "                ax.hist(DF_gene.loc[reg_genes].comp,alpha=0.7,bins=bins,\n",
    "                    range=(xmin,xmax), label=reg_label,color=reg_color,linewidth=0) \n",
    "        \n",
    "        # Plot multireg last\n",
    "        ax.hist(DF_gene.loc[multireg_genes].comp,alpha=0.5,bins=bins,\n",
    "            range=(xmin,xmax), label=multireg_label,color=multireg_color,linewidth=0)\n",
    "\n",
    "    \n",
    "    # Add vertical lines       \n",
    "    ymin,ymax=ax.get_ylim()\n",
    "    if len(tfs) > 0:\n",
    "        labelstr = 'P-value = {:.0e}\\nPrecision = {:.0f}%\\nRecall = {:.0f}%'.format(tf_info.pvalue, \\\n",
    "                              tf_info.precision*100,tf_info.recall*100)\n",
    "    else:\n",
    "        labelstr = None\n",
    "        \n",
    "    ax.vlines([ica_data.thresholds[k],-ica_data.thresholds[k]],0,1000,linestyles='dashed',\n",
    "               label=labelstr,linewidth=0.5)\n",
    "        \n",
    "    # Add gene names to each bin\n",
    "    eps = 1e-5\n",
    "    for x in np.arange(xmin,xmax,width):\n",
    "        bin_genes = DF_gene[(x < DF_gene.comp) & \n",
    "                    (DF_gene.comp <= x+width)]\n",
    "        # Only add names to bins outside the threshold\n",
    "        if (x <= -ica_data.thresholds[k]-eps or x >= ica_data.thresholds[k]-eps) and len(bin_genes) > 0:\n",
    "            y = max(bin_genes.color.value_counts())\n",
    "            y *= 1.1\n",
    "            \n",
    "            # If more than X genes, keep only largest X-1 and include \"+Y\"\n",
    "            max_genes1 = 9\n",
    "            max_genes2 = 16\n",
    "            if len(bin_genes) > max_genes2:\n",
    "                leftover = bin_genes.reindex(abs(bin_genes.comp).sort_values()[:-7].index)\n",
    "                bin_genes = bin_genes.reindex(abs(bin_genes.comp).sort_values()[-7:].index)\n",
    "                bin_genes.loc['Other','gene_name'] = '+{:d}'.format(len(leftover))\n",
    "                bin_genes.loc['Other','color'] = '#000000'\n",
    "            elif len(bin_genes) > max_genes1:\n",
    "                leftover = bin_genes.reindex(abs(bin_genes.comp).sort_values()[:-max_genes1].index)\n",
    "                bin_genes = bin_genes.reindex(abs(bin_genes.comp).sort_values()[-max_genes1:].index)\n",
    "                bin_genes.loc['Other','gene_name'] = '+{:d}'.format(len(leftover))\n",
    "                bin_genes.loc['Other','color'] = '#000000'\n",
    "            \n",
    "            # Draw text\n",
    "            for i,row in bin_genes.iterrows():\n",
    "                ax.text(x+width/2,y,'${}$'.format(row['gene_name']),color=colorscale(row.color,.75),\n",
    "                        ha='center',va='bottom',fontsize=7,fontweight='bold',fontstyle='italic')\n",
    "                y *= 1.65\n",
    "    \n",
    "    # Draw legend\n",
    "    legend = ax.legend(fontsize=7,frameon=True,ncol=(int(len(tfs)/3)+1))\n",
    "    legend.get_frame().set_facecolor('white')\n",
    "    legend.get_frame().set_linewidth(1)\n",
    "    legend.get_frame().set_alpha(1)\n",
    "    \n",
    "    # Set y axis tick size\n",
    "    ax.tick_params(axis='y',which='minor',length=3)\n",
    "    ax.tick_params(axis='y',which='major',length=5)\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fit_line(x,y,ax):\n",
    "    params,r2 = get_fit(x,y)\n",
    "    label = '$\\mathit{{R^2_{{adj}}}}$ = {:.2f}'.format(r2)\n",
    "    \n",
    "    plot_params = {'c':'gray','linewidth':1,'label':label}\n",
    "    \n",
    "    if len(params) == 2:\n",
    "        xlim = np.array([min(x),max(x)])\n",
    "        ax.plot(xlim,solid_line(xlim,*params),\n",
    "            **plot_params)\n",
    "    else:\n",
    "        xmin = min(x)\n",
    "        xmax = max(x)\n",
    "        mid = params[2]\n",
    "        xvals = np.array([xmin,mid,xmax])\n",
    "        ax.plot(xvals,broken_line(xvals,*params),\n",
    "            **plot_params)\n",
    "    return r2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def regulon_scatter(ica_data,k,tf,ax):\n",
    "    # Separate KO experiments\n",
    "    ko_cols = [exp for exp in ica_data.A.columns if 'del'+tf in exp \\\n",
    "               or 'del_'+tf in exp \\\n",
    "               or 'del'+tf.lower() in exp \\\n",
    "               or 'del_'+tf.lower() in exp]\n",
    "    \n",
    "    other_cols = set(ica_data.A.columns) - set(ko_cols) \n",
    "\n",
    "    \n",
    "    # Make scatter plot only if TF expression is in our dataset\n",
    "    if tf in ica_data.name2num.keys() and ica_data.name2num[tf] in ica_data.X.index:\n",
    "        ax.scatter(ica_data.X.loc[ica_data.name2num[tf],other_cols],\n",
    "                   ica_data.A.loc[k,other_cols],s=5,\n",
    "                   label='Conditions')\n",
    "\n",
    "        # Add colors to TF KO experiments\n",
    "        if len(ko_cols) > 0:\n",
    "            ax.scatter(ica_data.X.loc[ica_data.name2num[tf],ko_cols],\n",
    "                       ica_data.A.loc[k,ko_cols],s=5,\n",
    "                       label='{} KO'.format(tf))\n",
    "        \n",
    "        # Draw best fit line\n",
    "        r2 = fit_line(ica_data.X.loc[ica_data.name2num[tf],other_cols],\n",
    "                      ica_data.A.loc[k,other_cols],\n",
    "                      ax)\n",
    "\n",
    "        ax.set_ylabel('I-Modulon Activity',fontsize=10)\n",
    "        ax.set_xlabel(r'$\\mathit{{{}}}$ Expression Level'.format(tf),fontsize=10)\n",
    "        \n",
    "        # Draw and format legend\n",
    "        legend = ax.legend(frameon=True,fontsize=6)\n",
    "        legend.get_frame().set_facecolor('white')\n",
    "        legend.get_frame().set_linewidth(1)\n",
    "    else:\n",
    "        ax.axis('off')\n",
    "        \n",
    "    return ax,r2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def regulon_venn(ica_data,k,tf,ax=None):\n",
    "    if ica_data.regulon_mapping is None:\n",
    "        raise ValueError('No TRN information found')\n",
    "    \n",
    "    if ax == None:\n",
    "        fig,ax = plt.subplots(figsize=(5,5))\n",
    "    \n",
    "    # Take care of and/or enrichments\n",
    "    if '+' in tf:\n",
    "        reg_list = []\n",
    "        for tfx in tf.split('+'):\n",
    "            reg_list.append(set(ica_data.trn[ica_data.trn.TF==tfx].gene_id.unique()))\n",
    "        reg_genes = set.intersection(*reg_list)\n",
    "    elif '/' in tf:\n",
    "        reg_genes = set(ica_data.trn[ica_data.trn.TF.isin(tf.split('/'))].gene_id.unique())\n",
    "    else:\n",
    "        reg_genes = set(ica_data.trn[ica_data.trn.TF==tf].gene_id.unique())\n",
    "    \n",
    "    # Get component genes and operons\n",
    "    comp_genes = set(ica_data.show_enriched(k).index)\n",
    "    \n",
    "    reg_operons = len(ica_data.genes2operons(reg_genes-comp_genes))\n",
    "    comp_operons = len(ica_data.genes2operons(comp_genes-reg_genes))\n",
    "    both_operons = len(ica_data.genes2operons(reg_genes & comp_genes))\n",
    "    \n",
    "    # Draw venn diagram and resize texts\n",
    "    venn = venn2((reg_genes,comp_genes),\n",
    "                 set_labels=('Regulon\\nGenes','I-Modulon\\nGenes'),\n",
    "                 ax=ax)\n",
    "    for text in venn.set_labels:\n",
    "        text.set_fontsize(10)\n",
    "        if text.get_text() == u'Regulon\\nGenes':\n",
    "            text.set_color('darkred')\n",
    "        else:\n",
    "            text.set_color('darkgreen')\n",
    "    [reg_venn,comp_venn,both_venn] = venn.subset_labels\n",
    "\n",
    "    # Add operon numbers to labels\n",
    "    comp_venn.set_fontsize(10)\n",
    "    comp_venn.set_text(comp_venn.get_text()+'\\n({:d})'.format(comp_operons))\n",
    "    both_venn.set_fontsize(10)\n",
    "    both_venn.set_text(both_venn.get_text()+'\\n({:d})'.format(both_operons))\n",
    "    reg_venn.set_fontsize(10)\n",
    "    reg_venn.set_text(reg_venn.get_text()+'\\n({:d})'.format(reg_operons))\n",
    "\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Final PDF Maker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_axes(tfs,motif):\n",
    "    \n",
    "    # Set height ratios\n",
    "    height = [50,15,\n",
    "              20,15,\n",
    "              40,10]\n",
    "    \n",
    "    if len(tfs) > 0:\n",
    "        height += [40,10]\n",
    "    if motif:\n",
    "        height += [20,3]\n",
    "        \n",
    "    # Set TF scatterplots\n",
    "    scatter_tfs = []\n",
    "    for tf in tfs:\n",
    "        if (tf in ica_data.trn.TF.unique() and \\\n",
    "            tf in ica_data.name2num.keys() and \\\n",
    "            ica_data.name2num[tf] in ica_data.X.index):\n",
    "            scatter_tfs.append(tf)\n",
    "    \n",
    "    ## WARNING NOTE: HARDCODED EDGE CASE\n",
    "    if tfs == ['csqR']:\n",
    "        scatter_tfs = ['yihW']\n",
    "        \n",
    "    # Set width ratio\n",
    "    width = [5,2,80,10,40,40]\n",
    "    \n",
    "    # Set up gridspec\n",
    "    fig = plt.figure(figsize=(8.5,11))\n",
    "    gs = gridspec.GridSpec(len(height),len(width),\n",
    "                           height_ratios=height,width_ratios=width,\n",
    "                           left=0.08,right=0.92,top=0.92,bottom=0.05,\n",
    "                           wspace=0,hspace=0)\n",
    "    gs.update()\n",
    "    \n",
    "    # Populate axis dict\n",
    "    axes = {}\n",
    "    axes['s'] = plt.subplot(gs[0, 2])\n",
    "    axes['a'] = plt.subplot(gs[2, 1:])\n",
    "    axes['hist'] = plt.subplot(gs[4,1:])\n",
    "    \n",
    "    \n",
    "    if len(tfs) > 0:\n",
    "        if len(scatter_tfs) == 1:\n",
    "            axes['scatter'] = [(scatter_tfs[0],plt.subplot(gs[6,4:]))]\n",
    "        elif len(scatter_tfs) > 1:\n",
    "            ax2 = plt.subplot(gs[6,5])\n",
    "            ax2.set_ylabel('')\n",
    "            ax2.tick_params(axis='y',left=False,right=True,\n",
    "                            labelleft=False,labelright=True)\n",
    "            \n",
    "            axes['scatter'] = [(scatter_tfs[0],plt.subplot(gs[6,4])),\n",
    "                               (scatter_tfs[1],ax2)]\n",
    "        \n",
    "        axes['venn'] = plt.subplot(gs[6,1:3]) \n",
    "        \n",
    "        \n",
    "        if motif: \n",
    "            axes['motif'] = plt.subplot(gs[8,1:3])\n",
    "            axes['tomtom'] = plt.subplot(gs[8:,4:])\n",
    "\n",
    "    elif motif:\n",
    "            axes['motif'] = plt.subplot(gs[6,1:3])\n",
    "            axes['tomtom'] = plt.subplot(gs[6:,4:]) \n",
    "    \n",
    "    return fig,axes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "from scipy.optimize import OptimizeWarning\n",
    "warnings.simplefilter(action='ignore', category=OptimizeWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unchar_locs = [i for i,row in enrich.sort_values('name').iterrows() if 'uncharacterized' in row['name']]\n",
    "enrich_sort = enrich.loc[set(enrich.index) - set(unchar_locs)].sort_values('name')\n",
    "enrich_sort = pd.concat([enrich_sort,enrich.loc[unchar_locs]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "matplotlib.rcParams.update({'figure.max_open_warning': 0})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "r2_dict = {}\n",
    "\n",
    "with PdfPages(DATA_DIR+'modulons.pdf') as pdf:\n",
    "    for i,row in tqdm(enrich_sort.iterrows(),total=len(enrich)):\n",
    "        k = row['name']\n",
    "        # Get I-modulon Title\n",
    "        \n",
    "        if 'uncharacterized' in k:\n",
    "            title = 'Uncharacterized I-Modulon #{}'.format(k[-1])\n",
    "        else:\n",
    "            title = r'${}$ I-Modulon'.format(k)\n",
    "            \n",
    "        ########################\n",
    "        ## Get TF enrichments ##\n",
    "        ########################\n",
    "        \n",
    "        if pd.isnull(row.TF):\n",
    "            enrich_type = None\n",
    "            tfs = []\n",
    "        elif '+' in row.TF:\n",
    "            enrich_type = lambda x,y: x and y\n",
    "            tfs = row.TF.split('+')\n",
    "        elif '/' in row.TF:\n",
    "            enrich_type = lambda x,y: x or y\n",
    "            tfs = row.TF.split('/')\n",
    "        else:\n",
    "            enrich_type = None\n",
    "            tfs = [row.TF]\n",
    "        \n",
    "        # Check if there is an issue\n",
    "        if any([tf not in ica_data.trn.TF.unique() for tf in tfs]):\n",
    "            print('WARNING:',tf)\n",
    "            tfs = []\n",
    "        \n",
    "        ######################\n",
    "        ## Initialize Plots ##\n",
    "        ######################\n",
    "        \n",
    "        fig,axes = set_axes(tfs,k in DF_motif_final.index)\n",
    "\n",
    "        # Plot genes\n",
    "        plot_genes(ica_data,k,ax=axes['s'])\n",
    " \n",
    "        # Plot conditions\n",
    "        plot_samples_bar(ica_data,k,axes['a'])\n",
    "        \n",
    "        # Plot histogram\n",
    "        plot_histogram(ica_data,k,tfs,row,axes['hist'])\n",
    "        \n",
    "        # Plot venn diagram\n",
    "        if 'venn' in axes.keys():\n",
    "            regulon_venn(ica_data,k,row.TF,axes['venn'])\n",
    "\n",
    "        # Plot scatter\n",
    "        if 'scatter' in axes.keys():\n",
    "            ylims = []\n",
    "            for tf,ax in axes['scatter']:\n",
    "                _,r2 = regulon_scatter(ica_data,k,tf,ax)\n",
    "                r2_dict[(k,tf)] = r2\n",
    "                ylims.append(ax.get_ylim())\n",
    "            \n",
    "            # Ensure dual y-axes have same ylims\n",
    "            mins,maxes = zip(*ylims)\n",
    "            for _,ax in axes['scatter']:\n",
    "                ax.set_ylim(min(mins),max(maxes))\n",
    "        \n",
    "        # Plot motif\n",
    "        if 'motif' in axes.keys():\n",
    "            motif_ax = axes['motif']\n",
    "            tomtom_ax = axes['tomtom']\n",
    "            motif_ax.set_frame_on(False)\n",
    "            motif_ax.set_yticklabels([])\n",
    "            motif_ax.set_xticklabels([])\n",
    "            motif_ax.tick_params(bottom=False,left=False)\n",
    "            tomtom_ax.set_frame_on(False)\n",
    "            tomtom_ax.set_yticklabels([])\n",
    "            tomtom_ax.set_xticklabels([])\n",
    "            tomtom_ax.tick_params(bottom=False,left=False)\n",
    "\n",
    "            # Get motif\n",
    "            motif_row = DF_motif_final.loc[k]\n",
    "            im = plt.imread(motif_row.motif_img)\n",
    "            motif_ax.imshow(im)\n",
    "\n",
    "            xlabel_str = 'Motif E-value: {:.2e}\\nOperons with Upstream Motif: {:.0f}%'\n",
    "            xlabel = xlabel_str.format(motif_row.e_value,100*motif_row.motif_frac)\n",
    "            motif_ax.set_xlabel(xlabel,rotation='horizontal',fontsize=10,\n",
    "                          horizontalalignment='center',verticalalignment='top')            \n",
    "\n",
    "            # Plot tomtom output if relevant\n",
    "            if motif_row.tomtom_str != '':\n",
    "                im = plt.imread(motif_row.tomtom_img)\n",
    "                tomtom_ax.imshow(im)\n",
    "                tomtom_ax.set_xlabel(motif_row.tomtom_str,fontsize=10)\n",
    "            else:\n",
    "                tomtom_ax.axis('off')\n",
    "        \n",
    "        fig.suptitle(title,fontsize=26,fontweight='bold',va='baseline',y=0.96)\n",
    "        if pd.isnull(row.Regulator):\n",
    "            sub_str = ''\n",
    "        else:\n",
    "            sub_str = 'Regulated by: {}\\n'.format(row.Regulator)\n",
    "        if pd.notnull(row.Function):\n",
    "            sub_str += 'Biological Function: {}'.format(row.Function)\n",
    "        \n",
    "        fig.text(0.5,0.952,sub_str,ha='center',va='top',fontsize=12)\n",
    "        pdf.savefig(fig,dpi=1200)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
